joy_teleop/test/joy_teleop.test.py

```diff
@@ -0,0 +1,112 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+import time
+import unittest
+
+import launch
+import launch_ros
+import launch_ros.actions
+import rclpy
+import rclpy.context
+import rclpy.executors
+import sensor_msgs.msg
+import std_msgs.msg
+
+from ament_index_python.packages import get_package_share_directory
+
+
+def generate_test_description(ready_fn):
+    # Necessary to get real-time stdout from python processes:
+    proc_env = os.environ.copy()
+    proc_env['PYTHONUNBUFFERED'] = '1'
+
+    parameters_file = os.path.join(
+        get_package_share_directory('joy_teleop'),
+        'config', 'joy_teleop_example.yaml'
+    )
+
+    joy_teleop_node = launch_ros.actions.Node(
+        package='joy_teleop',
+        node_executable='joy_teleop',
+        parameters=[parameters_file],
+        env=proc_env,
+    )
+
+    return launch.LaunchDescription([
+            joy_teleop_node,
+            # Start tests right away - no need to wait for anything
+            launch.actions.OpaqueFunction(function=lambda context: ready_fn()),
+           ])
+
+
+class TestJoyTeleop(unittest.TestCase):
+
+    @classmethod
+    def setUpClass(cls):
+        cls.context = rclpy.context.Context()
+        rclpy.init(context=cls.context)
+        cls.node = rclpy.create_node('test_node', context=cls.context)
+
+        # The demo node listener has no synchronization to indicate when it's ready to start
+        # receiving messages on the /chatter topic.  This plumb_listener method will attempt
+        # to publish for a few seconds until it sees output
+        cls.publisher = cls.node.create_publisher(
+            sensor_msgs.msg.Joy,
+            'joy',
+            1
+        )
+        cls.msg = sensor_msgs.msg.Joy()
+
+    @classmethod
+    def tearDownClass(cls):
+        cls.node.destroy_node()
+
+    def spin_rclpy(self, timeout_sec):
+        executor = rclpy.executors.SingleThreadedExecutor(context=self.context)
+        executor.add_node(self.node)
+        try:
+            executor.spin_once(timeout_sec=timeout_sec)
+        finally:
+            executor.remove_node(self.node)
+
+    def test_topic_hello(self):
+
+        # Expect the talker to publish strings on '/chatter'
+        msgs_rx = []
+        sub = self.node.create_subscription(
+            std_msgs.msg.String,
+            'chatter',
+            lambda msg: msgs_rx.append(msg),
+            10
+        )
+        self.addCleanup(self.node.destroy_subscription, sub)
+
+        self.msg.buttons = [2]
+        self.publisher.publish(self.msg)
+
+        # Wait until the talker transmits messages over the ROS topic
+        end_time = time.time() + 10
+        while time.time() < end_time:
+            self.publisher.publish(self.msg)
+            self.spin_rclpy(1.0)
+            if len(msgs_rx) > 2:
+                break
+
+        self.assertGreater(len(msgs_rx), 1)
+
+        for msg in msgs_rx:
```

Comment:
First of all, `msgs_rx` is a terrible name in high level code. I also think this code is not really being used.

`cls.publisher` is never used, therefore no `joy_teleop` is being tested?